### PR TITLE
Replace jshint and jscs with eslint

### PR DESCRIPTION
**Pull request by jdforrester**
We've done this in pretty much all the main repos now. Time to get back to this one (replacing #107 from July 2016). Split into a number of small commits for easier reviewing.

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -21,27 +22,24 @@ module.exports = function ( grunt ) {
...
+				src: [
+					'<%= eslint.main %>'
+				]
 			},
-			main: {
-				src: '<%= jshint.all %>'
-			}
+			main: [
+				'*.js',
+				'src/**/*.js'
```
Comment (Nikerabbit) [inline Gruntfile.js]: It seems to me that excluding node_modules is needed to make the CI pass.

File: src/jquery.i18n.emitter.bidi.js [hunk 2]
```diff
@@ -45,6 +45,9 @@
 	 *
 	 * TODO: Does not handle BIDI control characters inside the text.
 	 * TODO: Does not handle unallocated characters.
+	 *
+	 * @param {String} text The text from which to extract initial directionality.
+	 * @returns {String} Directionality (either 'ltr' or 'rtl')
```
Comment (ricordisamoa) [inline src/jquery.i18n.emitter.bidi.js]: Why not `@return {string}`?

File: src/jquery.i18n.emitter.bidi.js [hunk 3]
```diff
@@ -70,6 +73,9 @@
 		 * strong directional codepoint" rule. Essentially, this works round the fact that
 		 * there is no embedding equivalent of U+2068 FSI (isolation with heuristic
 		 * direction inference). The latter is cleaner but still not widely supported.
+		 *
+		 * @param {String[]} nodes The text nodes from which to take the first item.
+		 * @returns {String} Wrapped String of content as needed.
```
Comment (ricordisamoa) [inline src/jquery.i18n.emitter.bidi.js]: Why not `@return {string}`?

File: src/jquery.i18n.language.js [hunk 4]
```diff
@@ -344,6 +343,7 @@
 		 *
 		 * @param {number} num Value to be converted
 		 * @param {boolean} integer Convert the return value to an integer
+		 * @returns {String} The number converted into a String.
```
Comment (ricordisamoa) [inline src/jquery.i18n.language.js]: Why not `@return {string}`?

File: src/jquery.i18n.emitter.bidi.js [hunk 5]
```diff
@@ -45,6 +45,9 @@
 	 *
 	 * TODO: Does not handle BIDI control characters inside the text.
 	 * TODO: Does not handle unallocated characters.
+	 *
+	 * @param {String} text The text from which to extract initial directionality.
```
Comment (ricordisamoa) [inline src/jquery.i18n.emitter.bidi.js]: Why not `{string}`?

File: src/jquery.i18n.emitter.bidi.js [hunk 6]
```diff
@@ -70,6 +73,9 @@
 		 * strong directional codepoint" rule. Essentially, this works round the fact that
 		 * there is no embedding equivalent of U+2068 FSI (isolation with heuristic
 		 * direction inference). The latter is cleaner but still not widely supported.
+		 *
+		 * @param {String[]} nodes The text nodes from which to take the first item.
```
Comment (ricordisamoa) [inline src/jquery.i18n.emitter.bidi.js]: Why not `{string[]}`?

**Comment by ricordisamoa** [comment]
`@return` (instead of `@returns`) and `string` instead of `String` seem to be more popular in the current codebase for Wikimedia JSDuck reasons.

File: Gruntfile.js (hunk 1 above)
Comment (jdforrester) [inline Gruntfile.js]: It shouldn't be needed, because it's not pulled in by this glob. CI passes locally.

Comment (ricordisamoa) [inline Gruntfile.js]: Try changing `'0.10'` to `'4'` in .travis.yml and it should work.

**Comment by jdforrester** [comment]
Gah. Rebasing on pushed branches seems impossible in GitHub? Meh.

**Pull request closed without merging**